stress/v2/statement/insert.go

```diff
@@ -0,0 +1,215 @@
+
+	// The Tracer prevents InsertStatement.Run() from returning early
+	Tracer *ponyExpress.Tracer
+
+	// Timestamp is #points to write and percision
+	Timestamp *Timestamp
+
+	// Templates turn into stringers
+	Templates Templates
+	stringers Stringers
+
+	// Number of series in this insert Statement
+	series int
+
+	// Returns the proper time for the next point
+	time func() int64
+
+	// Concurrency utiliities
+	sync.WaitGroup
+	sync.Mutex
+
+	// Timer for runtime and pps calculation
+	runtime time.Duration
+}
+
+func (i *InsertStatement) tags() map[string]string {
+	tags := map[string]string{
+		"number_fields":       i.numFields(),
+		"number_series":       fmtInt(i.series),
+		"number_points_write": fmtInt(i.Timestamp.Count),
+	}
+	return tags
+}
+
+// SetID statisfies the Statement Interface
+func (i *InsertStatement) SetID(s string) {
+	i.StatementID = s
+}
+
+// SetVars sets up the environment for InsertStatement to call it's Run function
+func (i *InsertStatement) SetVars(s *ponyExpress.StoreFront) chan<- string {
+	// Set the #series at 1 to start
+	i.series = 1
+
+	// Num series is the product of the cardinality of the tags
+	for _, tmpl := range i.Templates[0:i.TagCount] {
+		i.series *= tmpl.numSeries()
+	}
+
+	// make stringers from the templates
+	i.stringers = i.Templates.Init(i.series)
+
+	// Set the time function, keeps track of 'time' of the points being created
+	i.time = i.Timestamp.Time(s.StartDate, i.series, s.Precision)
+
+	// Set a commune on the StoreFront
+	s.Lock()
+	comCh := s.SetCommune(i.Name)
+	s.Unlock()
+
+	// Set the tracer
+	i.Tracer = ponyExpress.NewTracer(i.tags())
+
+	return comCh
+}
+
+// Run statisfies the Statement Interface
+func (i *InsertStatement) Run(s *ponyExpress.StoreFront) {
+
+	// Set variables on the InsertStatement and make the comCh
+	comCh := i.SetVars(s)
+
+	// TODO: Refactor to eleminate the ctr
+	// Start the counter
+	ctr := 0
+
+	// Create the first bytes buffer
+	buf := bytes.NewBuffer([]byte{})
+
+	runtime := time.Now()
+
+	for k := 0; k < i.Timestamp.Count; k++ {
+
+		// Increment the counter. ctr == k + 1?
+		ctr++
+
+		// Make the point from the template string and the stringers
+		point := fmt.Sprintf(i.TemplateString, i.stringers.Eval(i.time)...)
+
+		// Add the string to the buffer
+		buf.WriteString(point)
+		// Add a newline char to seperate the points
+		buf.WriteString("\n")
+
+		// If len(batch) == batchSize then send it
+		if ctr%s.BatchSize == 0 && ctr != 0 {
+			b := buf.Bytes()
+			// Trimming the trailing newline character
+			b = b[0 : len(b)-1]
+
+			// Create the package
+			p := ponyExpress.NewPackage(ponyExpress.Write, b, i.StatementID, i.Tracer)
+
+			// Use Tracer to wait for all operations to finish
+			i.Tracer.Add(1)
+
+			// Send the package
+			s.SendPackage(p)
+
+			// Reset the bytes Buffer
+			temp := bytes.NewBuffer([]byte{})
+			buf = temp
+		}
+
+		// Has to do with InsertStatement and QueryStatement communication
+		if len(comCh) < cap(comCh) {
```

Comment:
Seems redundant and racy to have a len/cap check wrapping a goroutine that attempts a non-blocking write.